firebase-installations/src/main/java/com/google/firebase/installations/remote/FirebaseInstallationServiceClient.java

```diff
@@ -97,46 +101,36 @@ public InstallationResponse createFirebaseInstallation(
+        httpsURLConnection.setDoOutput(true);
 
-      HttpsURLConnection httpsURLConnection = (HttpsURLConnection) url.openConnection();
-      httpsURLConnection.setConnectTimeout(NETWORK_TIMEOUT_MILLIS);
-      httpsURLConnection.setReadTimeout(NETWORK_TIMEOUT_MILLIS);
-      httpsURLConnection.setDoOutput(true);
-      httpsURLConnection.setRequestMethod("POST");
-      httpsURLConnection.addRequestProperty(CONTENT_TYPE_HEADER_KEY, JSON_CONTENT_TYPE);
-      httpsURLConnection.addRequestProperty(ACCEPT_HEADER_KEY, JSON_CONTENT_TYPE);
-      httpsURLConnection.addRequestProperty(CONTENT_ENCODING_HEADER_KEY, GZIP_CONTENT_ENCODING);
-      httpsURLConnection.addRequestProperty(X_ANDROID_PACKAGE_HEADER_KEY, context.getPackageName());
-      httpsURLConnection.addRequestProperty(
-          X_ANDROID_CERT_HEADER_KEY, getFingerprintHashForPackage());
-
-      GZIPOutputStream gzipOutputStream =
-          new GZIPOutputStream(httpsURLConnection.getOutputStream());
-      try {
-        gzipOutputStream.write(
-            buildCreateFirebaseInstallationRequestBody(fid, appId).toString().getBytes("UTF-8"));
-      } catch (JSONException e) {
-        throw new IllegalStateException(e);
-      } finally {
-        gzipOutputStream.close();
-      }
+        GZIPOutputStream gzipOutputStream =
+            new GZIPOutputStream(httpsURLConnection.getOutputStream());
+        try {
+          gzipOutputStream.write(
+              buildCreateFirebaseInstallationRequestBody(fid, appId).toString().getBytes("UTF-8"));
+        } catch (JSONException e) {
+          throw new IllegalStateException(e);
+        } finally {
+          gzipOutputStream.close();
+        }
 
-      int httpResponseCode = httpsURLConnection.getResponseCode();
-      switch (httpResponseCode) {
-        case 200:
-          return readCreateResponse(httpsURLConnection);
-        case 401:
-          throw new FirebaseInstallationServiceException(
-              UNAUTHORIZED_ERROR_MESSAGE, FirebaseInstallationServiceException.Status.UNAUTHORIZED);
-        default:
-          throw new FirebaseInstallationServiceException(
-              INTERNAL_SERVER_ERROR_MESSAGE,
-              FirebaseInstallationServiceException.Status.SERVER_ERROR);
+        int httpResponseCode = httpsURLConnection.getResponseCode();
+        switch (httpResponseCode) {
+          case 200:
+            return readCreateResponse(httpsURLConnection);
+          case 500:
+            retryCount++;
+            break;
+          default:
+            throw new FirebaseException(readErrorResponse(httpsURLConnection));
+        }
       }
+      throw new FirebaseException(INTERNAL_SERVER_ERROR_MESSAGE);
     } catch (IOException e) {
-      throw new FirebaseInstallationServiceException(
-          NETWORK_ERROR_MESSAGE + e.getMessage(),
-          FirebaseInstallationServiceException.Status.NETWORK_ERROR);
+      throw new FirebaseException(String.format(NETWORK_ERROR_MESSAGE, e.getMessage()));
```

Comment:
does it make sense to embed the IOException as a cause?

Reply:
IOException wraps many exceptions : MalformedException, ProtocolException, UnknownServiceException etc. My reasoning to add this as a cause for users to know what went wrong and handle accordingly.